# Data Inspection

About **ISWD**, & **oposum**

Using the scratch/notebook to make code and data in **official repository** more clear and transparent

In [1]:
import numpy as np
import joblib

In [2]:
%cd iswd/

C:\Project\group-1.3\ISWD\iswd


In [3]:
from DataHandler import *

# seed word weights

In [16]:
domain = 'bags_and_cases'
data_folder = f'../data/oposum/'

In [17]:
with open(data_folder + f'preprocessed/{domain.upper()}_word_mapping.txt', 'r') as f:
    word2id = {}
    id2word = {}
    for line in f:
        word, id = line.strip().split()
        id2word[int(id)] = word
        word2id[word] = int(id)

In [21]:
with open(data_folder + f'seed_words/{domain}.30-weights.txt', 'r') as f:
    aspects_ids = []
    seed_weights = []
    for line in f:
        seeds = []
        weights = []
        for tok in line.split():
            word, weight = tok.split(':')
            if word in word2id:
                seeds.append(word2id[word])
                weights.append(float(weight))
            else:
                print(f"word {word} not in vocab")
        aspects_ids.append(seeds)
        seed_weights.append(weights)

word 3080us not in vocab


In [26]:
from collections import Counter

In [34]:
Counter([id2word[a] for asp in aspects_ids for a in asp])

Counter({'pocket': 1,
         'cable': 2,
         'compartment': 1,
         'outside': 1,
         'lot': 2,
         'wallet': 1,
         'wish': 1,
         'connector': 1,
         'space': 2,
         'flap': 1,
         'charger': 1,
         'folder': 2,
         'pen': 1,
         'power': 1,
         'mouse': 1,
         'thing': 2,
         'inside': 2,
         'nice': 1,
         'small': 1,
         'daytimer': 2,
         'deforming': 1,
         'noise': 2,
         'headset': 1,
         'control': 2,
         'needing': 1,
         'earphone': 1,
         'huge': 1,
         'instead': 1,
         'accessory': 1,
         'outer': 1,
         'hassle': 2,
         'return': 2,
         'promptly': 2,
         'free': 2,
         'unable': 2,
         'difficult': 2,
         'arrived': 2,
         'christmas': 2,
         'replaced': 2,
         'easy': 2,
         'gift': 2,
         'returned': 2,
         'getting': 1,
         'amazon': 2,
         'week': 1,
  

sum for each aspect

In [23]:
for w in seed_weights:
    print(np.sum(w))

1.4850099999999997
9.45038
1.9780100000000005
1.4594199999999993
0.20390000000000005
3.885949999999999
2.571000000000001
0.9063899999999999
0.8045999999999998


# Data from official code

In [20]:
from DataHandler import *

In [ ]:
arg = {}

In [ ]:
ds = DataHandler()

## manually

In [10]:
import h5py

In [11]:
domain = 'bags_and_cases'
data_path = f'./data/oposum/preprocessed/{domain.upper()}.hdf5'

In [12]:
offi_train = []
offi_idx = []
f = h5py.File(data_path, 'r')
for b in f['data']:
    offi_train.extend(list(f['original/' + b][()]))
    offi_idx.extend(list(f['data/' + b][()]))

In [13]:
# offi_train
offi_train_str = list(map(lambda x: x.decode(), offi_train))


In [14]:
print(len(offi_train))
print(len(offi_idx))

588229
588229


In [16]:
# eg_str = "you 've"
eg_str = "when you 're"
eg_idx = offi_train_str.index(eg_str)
offi_idx[eg_idx]

array([9])

In [7]:
from nltk.corpus import stopwords
stw_list = set(stopwords.words('english'))
print(stw_list)
print("'ve" in stw_list)

{'aren', 'not', 'a', 'did', 'it', 't', 'yourselves', 'themselves', 'should', 'they', 'just', 'doesn', "don't", 'these', "you'd", 'mightn', "you're", 'the', 'into', 'that', 'own', 'his', 'again', 'how', 'haven', 'hasn', 've', 'who', 'after', "wouldn't", 'now', 'what', 'out', 'before', 'further', "needn't", 'shouldn', 'whom', 'ain', "couldn't", 'during', 'than', 'is', 'up', 'there', 'until', "doesn't", 'her', 'them', 'you', 'was', 'can', 'this', 'o', "mustn't", 'to', 'over', 'most', 'wouldn', 'd', 'above', "didn't", 'me', 'having', 'no', 'an', 'where', 'theirs', 'your', 'my', 'on', 'being', 'only', "should've", 'were', 'will', 'she', 'be', 'each', 'some', 'very', 'hadn', 'while', 'himself', 'does', 'from', 'herself', 'why', 'he', 'by', 'have', 'all', 'more', 're', 'doing', 'isn', "she's", 'am', 'in', 'between', 'him', "weren't", "that'll", 'then', 'myself', 'through', 'both', 'are', 'too', 'as', 'nor', 'here', "mightn't", 'against', "aren't", 'hers', 'under', 'shan', 'or', 'won', 'those'

In [ ]:
sorted(offi_idx, key=len)

In [19]:
# sorted(offi_train_str, key=len)
sorted(offi_train_str, key=len)[:10]
# sorted(offi_train_str, key=len)[-5:]

["'", 'e', 'C', '_', "'", '1', 'x', 'x', "'", '1']

# Data from unofficial code

In [10]:
import json
import numpy as np

In [42]:
domain = 'bags_and_cases'
data_path = f'../LeverageJustAFewKeywords/data/{domain}_train.json'

In [43]:
with open(data_path, 'r') as f:
    json_file = json.load(f)

In [44]:
# unoffi_train['original']
print(np.sum([len(batch) for batch in json_file['original']]))   # total number of segments
print(len(json_file['original']))    # number of batches

588229
11765


In [45]:
unoffi_train = [s for b in json_file['original'] for s in b]

In [49]:
sorted(unoffi_train, key=len)[-5:]

['that opens easily on top with a zipper at the TSA line , iPad , chargers , adapters , extra pens , passport , travel cards , meds , gum , wallet full of loyalty cards , folder of receipts , cough drops , tissues , other necessities , etc. under the flap in the roomy main area and in assorted pockets , and business cards , id badges , head set , memory sticks , sunglasses , and loose change on the outside pockets .',
 'that looks more for a passport than a card . a huge pocket for papers , and a hidden pocket for more papers or maybe a notebook , agenda , whatever . and of course , a pen holder for your pen . not a single pocket for memory cards like SDHC or CF , or maybe a pocket for an external CD/DVD-RW like my Samsung SE-S084 . anyway even with the pocket , will simply not fit for the lack of space so forget fill the card/papers slots',
 "I 'm able to carry a set of tools , a 50-pack spindle of CDR 's and DVDR 's , an array of hard-plastic double-CD/DVD cases containing various wo

# BERT embedding

statistics: in `bags_and_cases` training files, there are 588229 segments -> found that the code just take pre-computed BERT embedding from the data file

In [10]:
import joblib
import h5py

In [6]:
domain = "bags_and_cases"
method = "train"
bert_folder = "../data/pretrained_bert/oposum/"


In [7]:
bert_scodes = joblib.load(bert_folder + "{}_{}_scodes.pkl".format(domain, method))

In [9]:
len(bert_scodes)

588229

In [11]:
train_file_path = "../data/oposum/preprocessed/{}.hdf5".format(domain.upper())
train_file = h5py.File(train_file_path, 'r')

In [14]:
train_origial = []
for b in train_file['original']:
    train_origial.extend(train_file['original/' + b][()])
train_file.close()
len(train_origial)

588229

# word2vec

vocab length for bags_and_cases -> 30423

we use train/dev/test for building vocab, so the vocab size would be little larger

In [1]:
import h5py

In [6]:
all_domains = ['bags_and_cases', 'bluetooth', 'boots', 'keyboards', 'tv', 'vacuums']

In [7]:
for domain in all_domains:
    data_file = './data/oposum/preprocessed/' + domain.upper() + '.hdf5'
    f = h5py.File(data_file, 'r')
    wv = f['w2v'][()]
    print(f"domain {domain} vocab length: {len(wv)}\n")

domain bags_and_cases vocab length: 30423

domain bluetooth vocab length: 51229

domain boots vocab length: 30332

domain keyboards vocab length: 34066

domain tv vocab length: 59017

domain vacuums vocab length: 46228



# draft / scratch paper 

In [11]:
a = [[1, 2, 3, 4], [2, 3], [4, 5]]
b = np.array(a)
print(b)
c = np.array(a, dtype=object)
print(c)
print(type(c[0]))

[list([1, 2, 3, 4]) list([2, 3]) list([4, 5])]
[list([1, 2, 3, 4]) list([2, 3]) list([4, 5])]
<class 'list'>


ipykernel_launcher:2: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray


In [13]:
a = ['string1', 'a longer string', 'a much longer string']
b = np.array(a)
print(b)
print(type(b))

['string1' 'a longer string' 'a much longer string']
<class 'numpy.ndarray'>


In [14]:
a = np.arange(12)
b = a[3:4]
print(b)

[3]
